src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -121,63 +127,32 @@ static String getProperty(String propertyName, String def) {
                 // doing, we cannot promise success.
                 FileSystems.getDefault();
 
-                // First, try to find the library in our own jar
-                final boolean useExternalLib = Boolean.valueOf(getProperty("useExternalLib", "false"));
-                Exception loadingException = null;
-
-                String libraryName = System.mapLibraryName(LIBRARY_NAME);
-                if (useExternalLib) {
-                    loadingException = new RuntimeCryptoException("Skipping bundled library due to system property");
-                } else if (libraryName != null) {
-                    int index = libraryName.lastIndexOf('.');
-                    final String prefix = libraryName.substring(0, index);
-                    final String suffix = libraryName.substring(index, libraryName.length());
-
-                    final Path libPath = createTmpFile(prefix, suffix);
-
-                    try (final InputStream is = Loader.class.getResourceAsStream(libraryName);
-                         final OutputStream os = Files.newOutputStream(libPath, StandardOpenOption.CREATE,
-                                 StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)) {
-                        final byte[] buffer = new byte[16 * 1024];
-                        int read = is.read(buffer);
-                        while (read >= 0) {
-                            os.write(buffer, 0, read);
-                            read = is.read(buffer);
-                        }
-                        os.flush();
-                        System.load(libPath.toAbsolutePath().toString());
-                        return true;
-                    } catch (final Exception realException) {
-                        loadingException = realException;
-                    } catch (final Throwable realError) {
-                        loadingException = new RuntimeCryptoException("Unable to load native library", realError);
-                    } finally {
-                        Files.delete(libPath);
-                    }
-                } else {
-                    loadingException = new RuntimeCryptoException("Skipping bundled library null mapped name");
+                final Path libCryptoPath = writeResourceToTemporaryFile(System.mapLibraryName(LIBCRYPTO_NAME));
```

Comment:
why copy this to a temp file and then load from the temp file instead of just loading the original directly?

Reply:
You can only load libraries from actual files. This library is stored in our JAR and thus must be extracted and written to the filesystem before it can be loaded.

Reply:
got it. i was missing that this is reading directly from the JAR archive rather than unzipped contents on the filesystem.

Reply:
I wanted to make sure that I'm correctly understanding everything that is happening here, since this line is doing several things at once.

1. `System.mapLibraryName(LIBCRYPTO_NAME)` returns the platform specific resource name of the crypto library ACCP has a dependency on. (On Linux this would be `libcrypto.so`)
2. `writeResourceToTemporaryFile()` takes that library name, and calls Java's `Loader.class.getResourceAsStream()` which takes that resource name, and **looks inside the ACCP JAR file** for a resource with that name first.  
3. A `libcrypto.so` resource is found inside `AmazonCorrettoCryptoProvider.jar`, and is streamed to a temp file that is written to disk
4. The path to this temp file is returned, and is later passed down `loadLibCrypto()` where from C we load that crypto library into the current process.

Steps 2 and 3 were not immediately clear from reading just this line, and I wasn't sure if we were searching for currently installed resources on the platform (Eg searching the `/lib/x86_64-linux-gnu/` directory on Linux) or searching the resources inside the ACCP Jar file.